File: src/main/java/五毒/第三周/第二遍/permutation_ii.go
package main

func permuteUnique(nums []int) [][]int {
	result := make([][]int, 0)
	if len(nums) == 0 {
		return result
	}
	list := make([]int, 0)
	visited := make([]bool, len(nums))
	depth := 0
	back(&result, &list, &visited, depth, nums)
	return result
}

func back(result *[][]int, list *[]int, visited *[]bool, depth int, nums []int) {
	if depth == len(nums) {
		newList := make([]int, len(nums))
		copy(newList, *list)
		*result = append(*result, newList)
		return
	}
	var i int
	for i = 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] && (*visited)[i-1] == false {
			continue
		}
		if (*visited)[i] == false {
			(*visited)[i] = true
			*list = append(*list, nums[i])
			back(result, list, visited, depth+1, nums)
			(*visited)[i] = false
			*list = (*list)[0 : len(*list)-1]
		}
	}
}

File: src/main/java/五毒/第三周/第二遍/permutation_ii.java
package 五毒.第三周.第二遍;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class permutation_ii {
    public static List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if (nums.length == 0) {
            return result;
        }
        List<Integer> list = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        int depth = 0;
        Arrays.sort(nums);
        dfs(result, list, visited, depth, nums);
        return result;

    }

    private static void dfs(List<List<Integer>> result, List<Integer> list, boolean[] visited, int depth, int[] nums) {
        if (depth == nums.length) {
            result.add(new ArrayList<>(list));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1] && visited[i - 1] == false) {
                continue;
            }
            if (visited[i] == false) {
                visited[i] = true;
                list.add(nums[i]);
                dfs(result, list, visited, depth + 1, nums);
                visited[i] = false;
                list.remove(list.size() - 1);
            }
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 1, 2};
        System.out.println(permuteUnique(nums));
    }
}
